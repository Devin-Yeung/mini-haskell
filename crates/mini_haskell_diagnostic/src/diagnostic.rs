use crate::span::Span;

pub trait Diagnostic {
    fn span(&self) -> Span;
    fn message(&self) -> String;
}
