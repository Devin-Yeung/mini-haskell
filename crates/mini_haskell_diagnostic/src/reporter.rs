use crate::diagnostic::DiagnosticContext;
use ariadne::{ReportBuilder, Source};
use std::io::Cursor;
use std::ops::Range;

pub struct Reporter<'a> {
    // name of the source file
    // name: &'a str,
    // source code
    source: &'a str,
    // reporter config
    config: ReporterConfig,
    // diagnostic context
    context: DiagnosticContext,
}

impl<'a> Reporter<'a> {
    pub fn stringify(self) -> Vec<String> {
        self.context
            .reports
            .into_iter()
            .map(|report| {
                let mut builder: ReportBuilder<Range<usize>> =
                    ariadne::Report::build(ariadne::ReportKind::Error, (), report.offset)
                        .with_message(report.message)
                        .with_config(ariadne::Config::default().with_color(self.config.color));

                for label in report.labels {
                    builder = builder
                        .with_label(label.span.into())
                        .with_message(label.hint);
                }
                let report = builder.finish();

                let mut cursor = Cursor::new(Vec::<u8>::new());
                report
                    .write(ariadne::Source::from(self.source), &mut cursor)
                    .unwrap();
                String::from_utf8(cursor.into_inner()).unwrap()
            })
            .collect::<Vec<_>>()
    }
}

pub struct ReporterBuilder<'a> {
    source: Option<&'a str>,
    config: Option<ReporterConfig>,
    context: Option<DiagnosticContext>,
}

impl<'a> ReporterBuilder<'a> {
    pub fn new() -> Self {
        Self {
            source: None,
            config: None,
            context: None,
        }
    }

    pub fn source(mut self, source: &'a str) -> Self {
        self.source = Some(source);
        self
    }

    pub fn context(mut self, context: DiagnosticContext) -> Self {
        self.context = Some(context);
        self
    }

    pub fn config(mut self, config: ReporterConfig) -> Self {
        self.config = Some(config);
        self
    }
}

pub struct ReporterConfig {
    color: bool,
}
