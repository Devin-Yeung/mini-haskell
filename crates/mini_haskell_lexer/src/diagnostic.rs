#[cfg(test)]
mod test {
    use crate::error::LexingError;
    use crate::lexer::Token;
    use mini_haskell_diagnostic::diagnostic::diagnostic;
    use testsuite::unittest;

    unittest!(invalid_token, |path, _| {
        let result = diagnostic(path, |src| {
            Token::tokens(src)
                .into_iter()
                .filter(|tok| tok.is_err())
                .map(|tok| tok.err().unwrap())
                .collect::<Vec<LexingError>>()
        });

        insta::assert_snapshot!(result);
    });
}
