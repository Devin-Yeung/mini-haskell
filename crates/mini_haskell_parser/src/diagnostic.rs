#[cfg(test)]
mod test {
    use crate::parser::Parser;
    use mini_haskell_diagnostic::diagnostic::diagnostic;
    use testsuite::unittest;

    unittest!(unexpected_token, |path, _| {
        let result = diagnostic(path, |src| {
            let (_, errors) = Parser::parse(src);
            errors
        });

        insta::assert_snapshot!(result);
    });
}
