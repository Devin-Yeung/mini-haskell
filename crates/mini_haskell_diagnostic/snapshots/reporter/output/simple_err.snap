---
source: crates/mini_haskell_diagnostic/src/reporter.rs
expression: result
---
  × This is an example error
   ╭─[snapshots/reporter/input/simple_err.hs:1:1]
 1 │ Hello World!
   ·      ▲
   ·      ╰── Remove this space
   ╰────
  help: This is an example help msg

